packages/streams-adapter/server/server.go

```diff
@@ -0,0 +1,431 @@
+		})
+		return
+	}
+
+	// Canonicalize and filter request parameters once using the alias index, including overrides.
+	canonicalParams, err := helpers.BuildCacheKeyParams(reqData.Data)
+	if err != nil {
+		s.logger.Error("Failed to canonicalize request params", "error", err, "requestData", reqData.Data)
+
+		// If we cannot canonicalize the request (e.g. unknown/unsupported endpoint),
+		// do not attempt to use the cache or subscribe and return a 500 to the caller.
+		errorResp := errorResponsePool.Get().(*ErrorResponseData)
+		defer errorResponsePool.Put(errorResp)
+
+		errorResp.Error.Name = "AdapterError"
+		errorResp.Error.Message = "Unable to subsribe to an asset pair with the the requested data"
+
+		c.JSON(http.StatusInternalServerError, errorResp)
+		return
+	}
+
+	// Set request params for metrics tracking
+	middleware.SetRequestParams(c, canonicalParams)
+
+	// Check if cached response exists
+	cachedResponse := s.cache.Get(canonicalParams)
+
+	// Check if observation exists in cache
+	if cachedResponse == nil {
+		subscriptionKey, err := helpers.CalculateCacheKey(canonicalParams)
+		if err != nil {
+			s.logger.Error("Failed to calculate cache key", "error", err, "requestParams", canonicalParams)
+			// If we cannot calculate a subscription key, do not attempt to subscribe
+			// and return a 500 to the caller.
+			errorResp := errorResponsePool.Get().(*ErrorResponseData)
+			defer errorResponsePool.Put(errorResp)
+
+			errorResp.Error.Name = "AdapterError"
+			errorResp.Error.Message = "Unable to subsribe to an asset pair with the the requested data"
+
+			c.JSON(http.StatusInternalServerError, errorResp)
+			return
+		}
+
+		// Only subscribe if we're not already subscribing to this request
+		// This prevents thundering herd of duplicate subscription requests
+		if _, alreadySubscribing := s.subscriptionTracker.LoadOrStore(subscriptionKey, true); !alreadySubscribing {
+			s.logger.Debug("Initiating new subscription", "requestParams", canonicalParams, "subscriptionKey", subscriptionKey)
+			go func(key string, params types.RequestParams) {
+				s.subscribeToAsset(params)
+				// Remove from tracker after subscription attempt completes
+				// Allow retries after 10 seconds if data still not available
+				time.Sleep(10 * time.Second)
+				s.subscriptionTracker.Delete(key)
+			}(subscriptionKey, canonicalParams)
+		} else if s.config.LogLevel == "debug" {
+			s.logger.Debug("Subscription already in progress, skipping", "key", subscriptionKey)
+		}
+
+		// Get error response from pool
+		errorResp := errorResponsePool.Get().(*ErrorResponseData)
+		defer errorResponsePool.Put(errorResp)
+
+		errorResp.Error.Name = "AdapterError"
+		errorResp.Error.Message = "The EA has not received any values from the Data Provider for the requested data yet. Retry after a short delay, and if the problem persists raise this issue in the relevant channels."
+
+		c.JSON(http.StatusGatewayTimeout, errorResp)
+		return
```

Comment:
This block of code is supposed to execute when the attempt to subscribe times out. The way the code is right now, it will fire off every time.

If we want this call to be blocking then we'll need some channel or wait group to sync the execution of the sub goroutine (line 291) and either return the requested value from the cache or return a timeout error.